resources/aws_ec2.go

```diff
@@ -174,22 +257,95 @@ func (obj *AwsEc2Res) Init() error {
 	obj.awsChan = make(chan *chanStruct)
 	obj.closeChan = make(chan struct{})
 	obj.wg = &sync.WaitGroup{}
+	obj.awsChan = make(chan *chanStruct)
+	obj.closeChan = make(chan struct{})
+
+	if obj.WatchListenAddr != "" {
+		// make sns client
+		snsSess, err := session.NewSession(&aws.Config{
+			Region: aws.String(obj.Region),
+		})
+		if err != nil {
+			return errwrap.Wrapf(err, "error creating sns session")
+		}
+		obj.snsClient = sns.New(snsSess)
+	}
 
 	return obj.BaseRes.Init() // call base init, b/c we're overriding
 }
 
 // Watch is the primary listener for this resource and it outputs events.
 func (obj *AwsEc2Res) Watch() error {
+	if obj.WatchEndpoint != "" {
+		return obj.snsWatch()
+	}
 	return obj.longpollWatch()
 }
 
+// snsWatch uses amazon cloudwatch events and simple notification service to
+// detect ec2 instance state changes.
+func (obj *AwsEc2Res) snsWatch() error {
+	send := false
+	var exit *error
+	var err error
+	if err = obj.snsMakeTopic(); err != nil {
```

Comment:
@purpleidea: This is clearly a more sensible pattern. Not sure what I was thinking.

Reply:
If all of these functions have side effects, eg they are reading or setting some value in the struct, then it's much harder to follow what's happening. In this case, if it's not needed that the variable be in the struct, have the values get passed in as arguments, and output and then passed to the next function.

In some cases, it's even fine to get the output, and then in this visible code store it in the struct after the function ends.

Reply:
That makes sense. Gotta get back into the haskell/functional programming mindset. Thanks